### PR TITLE
refactor(budget-optimizer): collapse default_constraints into constraints

### LINKED ISSUE #1416: Simplify optimizer parameter `default_constraints`
The optimizer use `custom_constraints` and `default_constraints` this can be simplified, and have all into `custom_constraints` as proposed by @wd60622 in #1358

### PR DESCRIPTION
**Pull request by anevolbap**
Closes #1416. See the issue thread for design discussion. Notebook updates land in a follow-up PR.

## Summary
- Rename `custom_constraints` to `constraints` on `BudgetOptimizer`.
- Remove `default_constraints`. Empty `constraints` auto-adds the default sum constraint; non-empty means the caller is in charge. Pass `build_default_sum_constraint()` explicitly to keep the default alongside customs.
- Remove `default_constraints` from `MMM.optimize_budget`.
- Simplify `set_constraints` to `set_constraints(constraints)`.

Hard removal, no deprecation cycle: v1.0 is the next release and carries breaking changes anyway (per @juanitorduz's review).

**BREAKING CHANGE**: `custom_constraints` and `default_constraints` are removed. Use `constraints` instead.

## Test plan
- [x] `tests/mmm/test_budget_optimizer.py` (22 pass)
- [x] `tests/mmm/test_budget_optimizer_mmm.py` + `tests/mmm/test_cost_per_unit.py` (77 pass)
- [x] pre-commit clean
- [ ] Notebook updates (follow-up PR)

### CODE REVIEW

File: pymc_marketing/mmm/budget_optimizer.py [hunk 1]
```diff
@@ -708,6 +717,72 @@ class BudgetOptimizer(BaseModel):
     _constraints: dict = PrivateAttr()
     _compiled_constraints: list[dict] = PrivateAttr()
 
+    @model_validator(mode="before")
```
Comment (anevolbap) [inline pymc_marketing/mmm/budget_optimizer.py]: Outdated: this validator was removed in 61d1a55c (hard removal, no deprecation shim).

**Review by anevolbap — COMMENTED** [review]
Adding a few inline notes to ease review.

File: pymc_marketing/mmm/budget_optimizer.py [hunk 2]
```diff
@@ -708,6 +717,72 @@ class BudgetOptimizer(BaseModel):
...
+
+        if "constraints" in data:
+            # Pydantic v2 wraps `ValueError` raised from a `mode="before"`
+            # validator into `pydantic.ValidationError` (itself a `ValueError`).
+            # `TypeError` is re-raised as-is, which is what we want here:
+            # passing mutually exclusive kwargs is a programming error, not a
+            # value error.
+            raise TypeError(
+                "Pass either `constraints` or the deprecated "
+                "`custom_constraints`/`default_constraints`, not both."
```
Comment (anevolbap) [inline pymc_marketing/mmm/budget_optimizer.py]: Outdated: removed in 61d1a55c.

File: pymc_marketing/mmm/budget_optimizer.py [hunk 3]
```diff
@@ -828,16 +903,39 @@ def model_post_init(self, context: Any, /) -> None:
 
         # 9. Build constraints
         self._constraints = {}
-        self.set_constraints(
-            default=self.default_constraints, constraints=self.custom_constraints
-        )
+        self.set_constraints(constraints=self.constraints)
 
     def set_constraints(self, constraints, default=None) -> None:
```
Comment (anevolbap) [inline pymc_marketing/mmm/budget_optimizer.py]: Outdated: `set_constraints` no longer takes a `default` kwarg after 61d1a55c; it is now `set_constraints(constraints)`.

File: pymc_marketing/mmm/mmm.py [hunk 4]
```diff
@@ -3796,18 +3803,25 @@ def optimize_budget(
...
             response_variable=response_variable,
-            custom_constraints=constraints,
-            default_constraints=default_constraints,
             budgets_to_optimize=budgets_to_optimize,
             budget_distribution_over_period=budget_distribution_over_period,
             cost_per_unit=cost_per_unit_da,
             model=self,
             compile_kwargs=self.compile_kwargs,
         )
+        if default_constraints is None:
```
Comment (anevolbap) [inline pymc_marketing/mmm/mmm.py]: Outdated: the legacy routing was removed in 61d1a55c; `optimize_budget` now forwards `constraints` directly.

**Comment by juanitorduz** [comment]
@carlosagostini  :)

File: pymc_marketing/mmm/budget_optimizer.py [hunk 5]
```diff
@@ -708,6 +717,72 @@ class BudgetOptimizer(BaseModel):
...
 
+    @model_validator(mode="before")
+    @classmethod
+    def _migrate_legacy_constraint_kwargs(cls, data: Any) -> Any:
+        """Translate deprecated ``custom_constraints``/``default_constraints`` to ``constraints``.
+
+        .. deprecated:: 0.20.0
+            ``custom_constraints`` and ``default_constraints`` are deprecated
+            and will be removed in 0.21.0. Use ``constraints`` instead.
+        """
```
Comment (juanitorduz) [inline pymc_marketing/mmm/budget_optimizer.py]: as we are migrating to v1.0 in the next release shall we simply spik this and change the name?

File: pymc_marketing/mmm/budget_optimizer.py [hunk 6]
```diff
@@ -708,6 +717,72 @@ class BudgetOptimizer(BaseModel):
...
+            raise TypeError(
+                "Pass either `constraints` or the deprecated "
+                "`custom_constraints`/`default_constraints`, not both."
+            )
+
+        custom = data.pop("custom_constraints", ())
+        # `default_constraints` defaulted to True under the old API.
+        default_flag = data.pop("default_constraints", True)
+
+        warnings.warn(
```
Comment (juanitorduz) [inline pymc_marketing/mmm/budget_optimizer.py]: same here

**Review by juanitorduz — COMMENTED** [review]
@cetagostini can we simply rename and remove the deprecation warnings are we are planning the v1 as our next release (there will be many breaking changes anyway)

**Comment by anevolbap** [comment]
Done in 61d1a55c. Dropped the deprecation shim and the legacy kwargs entirely, only `constraints` remains.

The deprecation-shim version is the previous commit (c5deff55), so if @cetagostini still prefers a one-release `DeprecationWarning` we can revert to it quickly.

File: pymc_marketing/mmm/budget_optimizer.py (hunk 5 above)
Comment (anevolbap) [reply pymc_marketing/mmm/budget_optimizer.py]: Done in 61d1a55c. Dropped the validator and the legacy kwargs, only `constraints` remains.

File: pymc_marketing/mmm/budget_optimizer.py (hunk 6 above)
Comment (anevolbap) [reply pymc_marketing/mmm/budget_optimizer.py]: Removed in 61d1a55c along with the rest of the shim.

**Comment by cetagostini** [comment]
Hey guys, coming back from a long time off. I'll be reviewing this week!

File: pymc_marketing/mmm/budget_optimizer.py [hunk 7]
```diff
@@ -613,10 +613,11 @@ class BudgetOptimizer(BaseModel):
...
-        Custom constraints for the optimizer.
-    default_constraints : bool, optional
-        Whether to add a default sum constraint on the total budget. Default is True.
+    constraints : Sequence[Constraint], optional
+        Constraints for the optimizer. If empty, a default sum-equals-total-budget
+        constraint is added automatically. If non-empty, the caller is in charge:
+        no default is added. Pass ``build_default_sum_constraint()`` explicitly
+        to keep the sum constraint alongside custom ones.
     budget_distribution_over_period : xarray.DataArray, optional
         Fixed temporal distribution of each budget cell across periods.
```
Comment (cetagostini) [inline pymc_marketing/mmm/budget_optimizer.py:622]: **Missing type annotation** (project standard: "All function arguments and return types must be typed").

```python
# Current
def set_constraints(self, constraints) -> None:

# Suggested
def set_constraints(self, constraints: Sequence[Constraint]) -> None:
```

File: pymc_marketing/mmm/budget_optimizer.py [hunk 8]
```diff
@@ -658,14 +659,15 @@ class BudgetOptimizer(BaseModel):
         description="Mask defining a subset of budgets to optimize. Non-optimized budgets remain fixed at 0.",
     )
 
-    custom_constraints: Sequence[Constraint] = Field(
+    constraints: Sequence[Constraint] = Field(
         default=(),
```
Comment (cetagostini) [inline pymc_marketing/mmm/budget_optimizer.py:663]: **Redundant Constraint copy with dead defensive guard.**

The `Constraint` class types `constraint_type` as `Literal["eq", "ineq"]` — it should never be `None` unless the caller bypasses type checking. The `if c.constraint_type else "eq"` fallback is dead code.

The entire loop can be replaced with:
```python
self._constraints = {c.key: c for c in constraints}
```

If you want to keep the `None` guard for runtime safety, update the `Constraint` class to declare it explicitly (`constraint_type: Literal["eq", "ineq"] | None = "eq"`) so the intent is documented.

File: pymc_marketing/mmm/budget_optimizer.py [hunk 9]
```diff
@@ -658,14 +659,15 @@ class BudgetOptimizer(BaseModel):
...
-        description="Custom constraints for the optimizer.",
```
Comment (cetagostini) [inline pymc_marketing/mmm/budget_optimizer.py:663]: **Silent duplicate-key overwrite.** If two constraints share the same `key`, the second silently clobbers the first. Consider adding a guard:

```python
for c in constraints:
    if c.key in self._constraints:
        raise ValueError(
            f"Duplicate constraint key {c.key!r}. Keys must be unique."
        )
    self._constraints[c.key] = c
```

File: pymc_marketing/mmm/mmm.py [hunk 10]
```diff
@@ -4024,7 +4024,6 @@ def optimize_budget(
         response_variable: str = "total_media_contribution_original_scale",
         utility_function: UtilityFunctionType = average_response,
         constraints: Sequence[Constraint] = (),
```
Comment (cetagostini) [inline pymc_marketing/mmm/mmm.py:4026]: **Docstring: missing default value.** The `constraints` parameter doesn't mention its default `()`. For consistency with the rest of the docstring:

```
constraints : Sequence[Constraint], optional
    Constraints for the optimizer. Default is ``()`` (empty tuple).
```

File: tests/mmm/test_budget_optimizer.py [hunk 11]
```diff
@@ -268,13 +268,11 @@ def test_allocate_budget(
     optimizes ``total_media_contribution_original_scale`` (media-only, original
     units), so the expected allocation/response values intentionally differ.
     """
-    match = "Using default equality constraint"
-    with pytest.warns(UserWarning, match=match):
-        optimizer = BudgetOptimizer(
-            model=mmm_wrapper,
-            num_periods=30,
```
Comment (cetagostini) [inline tests/mmm/test_budget_optimizer.py:275]: **Test gap: no re-entrant `set_constraints` test.** All tests construct a fresh `BudgetOptimizer` and let `model_post_init` call `set_constraints`. There's no test that calls `optimizer.set_constraints(...)` on an already-initialized optimizer to verify re-entrant behavior (clearing old constraints, re-compilation).

Also missing: a test that `constraints=[]` (explicit empty list) behaves identically to the default `constraints=()`.

**Review by cetagostini — COMMENTED** [review]
## AI-Assisted Code Review

> This review was generated with AI assistance (opencode / mimo-v2.5-pro) to identify potential improvements. A human reviewer should validate each suggestion before applying.

---

### Summary

The refactor is clean and well-scoped. The API simplification (`custom_constraints` + `default_constraints` → `constraints`) is a clear improvement. Tests pass (22/22). Below are actionable suggestions to tighten the implementation.

### Issues Found

1. **Missing type annotation** on `set_constraints` parameter — violates project coding standards.
2. **Redundant Constraint copy** with a dead defensive `None` guard — can be simplified.
3. **Silent duplicate-key overwrite** — two constraints with the same key silently clobber each other.
4. **Test gap** — no test for re-entrant `set_constraints` calls, no test for explicit `constraints=[]` vs default.
5. **`optimize_budget` docstring** — `constraints` parameter doesn't mention the default value.

See inline comments below.

File: pymc_marketing/mmm/budget_optimizer.py (hunk 7 above)
Comment (anevolbap) [reply pymc_marketing/mmm/budget_optimizer.py:622]: Done in 34d2155a: `set_constraints(self, constraints: Sequence[Constraint])`.

File: pymc_marketing/mmm/budget_optimizer.py (hunk 8 above)
Comment (anevolbap) [reply pymc_marketing/mmm/budget_optimizer.py:663]: Applied in 34d2155a: store the `Constraint` objects directly (`self._constraints[c.key] = c`), dropping the per-item copy.

One correction though: the `c.constraint_type if c.constraint_type else "eq"` fallback was not strictly dead. `Constraint.__init__` does not default `constraint_type` (it is a required positional), so a `Constraint(constraint_type=None)` is constructible and the old loop normalized that falsy value to "eq". The `Literal["eq", "ineq"]` annotation is not enforced at runtime. In practice every builder (including `build_default_sum_constraint`) passes a valid "eq"/"ineq", so dropping the normalization is safe, but it was guarding against loosely constructed Constraints rather than being plain dead code.

Note this loop is unchanged from main, so it is out of scope for the rename.

File: pymc_marketing/mmm/budget_optimizer.py (hunk 9 above)
Comment (anevolbap) [reply pymc_marketing/mmm/budget_optimizer.py:663]: Done in 34d2155a, also out of scope for the rename but applied: duplicate keys now raise `ValueError` instead of silently overwriting. Covered by `test_duplicate_constraint_keys_raise`.

File: pymc_marketing/mmm/mmm.py (hunk 10 above)
Comment (anevolbap) [reply pymc_marketing/mmm/mmm.py:4026]: Done in 34d2155a: folded the `()` default into the `constraints` description.

File: tests/mmm/test_budget_optimizer.py (hunk 11 above)
Comment (anevolbap) [reply tests/mmm/test_budget_optimizer.py:275]: Done in 34d2155a: added `test_set_constraints_is_reentrant` (clears prior state and recompiles) and `test_constraints_empty_list_matches_default` (explicit `[]` matches the default `()`).

**Comment by anevolbap** [comment]
@cetagostini addressed all five points in 34d2155a (type annotation, direct-store loop, duplicate-key guard, `()` default in the docstring, plus re-entrant / empty-list / duplicate-key tests). One note: the `else "eq"` fallback was normalization rather than dead code (details in the thread), but dropping it is safe. The `build source distribution` failures are the `pymc-extras`/`arviz` pip conflict from #2616, unrelated to this branch. Ready for another look.

**Comment by cetagostini** [comment]
Hey @anevolbap great work!

I don't see anything more blocking here!

**Review by cetagostini — APPROVED** [review]

**Comment by cetagostini** [comment]
cc: @williambdean and @juanitorduz

**Comment by juanitorduz** [comment]
Thanks @anevolbap !
